CorsixTH/Src/th_gfx.cpp

```diff
@@ -1033,88 +1092,119 @@ inline void THChunkRenderer::_incrementPosition(int npixels)
+        {
+            // The value represents the length of the buffer to copy
+            size_t amt = inputVal;
+            if (dataLen < amt)
+                amt = dataLen;
+            chunkCopy(amt, inputData);
+            inputData += amt;
+            dataLen -= amt;
+        }
+        else
+        {
+            // Fill the next (256 - value) pixels with all bits set
+            // as the value is greater than 128 because of the above check it will
+            // be the next 1-128 pixels
+            chunkFill(256 - inputVal, 0xFF);
         }
     }
-    else
+}
+
+void THChunkRenderer::decodeChunksComplex(const uint8_t * inputData, size_t dataLen)
+{
+    while (!isEndOfBuffer() && dataLen > 0)
     {
-        while(!_isDone() && datalen > 0)
+        uint8_t inputVal = *inputData;
+        --dataLen;
+        ++inputData;
+
+        if (inputVal == 0)
         {
-            uint8_t b = *data;
-            --datalen;
-            ++data;
-            if(b == 0)
-            {
-                chunkFillToEndOfLine(0xFF);
-            }
-            else if(b < 0x80)
+            // If the value is 0 fill to the end of the line with all bits set
+            chunkFillToEndOfLine(0xFF);
+        }
+        else if (inputVal < 64)
+        {
+            // Copy the number of pixels represented by inputVal.
+            size_t amt = inputVal;
+            if (dataLen < amt)
+                amt = dataLen;
+            chunkCopy(amt, inputData);
+            inputData += amt;
+            dataLen -= amt;
+        }
+        // If the input has bit 128 set without bit 64
+        // i.e. in the range 128-191 
+        else if ((inputVal & 0xC0) == 0x80)
+        {
+            // Fill the with 0xFF for the number of pixels - 128
+            chunkFill(inputVal - 0x80, 0xFF);
+        }
+        else
+        {
+            // B is between 64 - 127 or 192 - 255
+            int amt;
+            uint8_t colour = 0;
+            // If inputVal is equal to 255 (0xFF)
+            if (inputVal == 0xFF)
             {
-                int amt = b;
-                if(datalen < amt)
-                    amt = datalen;
-                chunkCopy(amt, data);
-                data += amt;
-                datalen -= amt;
+                if (dataLen < 2)
+                {
+                    break;
+                }
+                // Copy the number of pixels given with the colour. 
+                amt = static_cast<int>(inputData[0]);
+                colour = inputData[1];
+                inputData += 2;
+                dataLen -= 2;
             }
             else
             {
-                chunkFill(0x100 - b, 0xFF);
+                // Bitwise And inputVal with 128, divide this by two
+                // then subtract this from inputVal. Then subtract
+                // 60 from inputVal. This gives the amount of pixels
+                // to copy.
+                amt = inputVal - 60 - (inputVal & 0x80) / 2;
```

Comment:
Comment that repeats what the code says is less than useless. At best, it actually repeats it (giving no new information), less optimal if it says something else (which can only give confusion). Instead, explain what is happening here at a higher level, and clarify the 60, as it's pure magic now.